Crash with invalid mapping

```

2018/11/01 20:37:26 events.go:125: AlertManager configured at http://alertmanager.sa-telemetry.svc:9093/api/v1/alerts
--
  | 2018/11/01 20:37:26 events.go:131: API available at localhost:9999
  | 2018/11/01 20:37:26 events.go:137: AMQP1.0 Publish address at http://messaging-internal-blue.sa-telemetry.svc:5672/collectd/alerts2
  | 2018/11/01 20:37:26 events.go:99: Debug:Config saconfig.EventConfiguration{Debug:true, AMQP1EventURL:"messaging-internal-blue.sa-telemetry.svc:5672/collectd/notify", ElasticHostURL:"http://elasticsearch.sa-telemetry.svc:9200 ", API:saconfig.EventAPIConfig{APIEndpointURL:"localhost:9999", AMQP1PublishURL:"http://messaging-internal-blue.sa-telemetry.svc:5672/collectd/alerts2 "}, AlertManagerURL:"http://alertmanager.sa-telemetry.svc:9093/api/v1/alerts ", AlertManagerEnabled:true, APIEnabled:true, PublishEventEnabled:true, ResetIndex:false, Prefetch:0, UniqueName:"", ServiceType:"events", IgnoreString:""}
  | 2018/11/01 20:37:26 events.go:152: Connecting to AMQP1 : amqp://messaging-internal-blue.sa-telemetry.svc:5672/collectd/notify
  | 2018/11/01 20:37:26 events.go:157: Listening.....
  | 2018/11/01 20:37:26 events.go:159: Connecting to ElasticSearch : http://elasticsearch.sa-telemetry.svc:9200
  | 2018/11/01 20:37:26 receiver.go:127: Parsing amqp://messaging-internal-blue.sa-telemetry.svc:5672/collectd/notify
  | 2018/11/01 20:37:26 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:26 client.go:60: Debug:ElasticSearch client created.
  | 2018/11/01 20:37:26 events.go:189: APIEndpoint server ready at : localhost:9999
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:27 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:27 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:28 events.go:194: Ready....
  | 2018/11/01 20:37:28 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:28 receiver.go:127: Debug: Getting message from AMQP[{"values":[0],"dstypes":["gauge"],"dsnames":["value"],"time":1541104650.005,"interval":10.000,"host":"nfvha-strg-03.oot.lab.eng.bos.redhat.com","plugin":"df","plugin_instance":"root","type":"df_complex","type_instance":"reserved"}]
  | 2018/11/01 20:37:28 receiver.go:127: Debug: Sending message to Notifier channel
  | 2018/11/01 20:37:28 receiver.go:127: Debug: Getting message from AMQP[{"values":[49611153408],"dstypes":["gauge"],"dsnames":["value"],"time":1541104650.005,"interval":10.000,"host":"nfvha-strg-03.oot.lab.eng.bos.redhat.com","plugin":"df","plugin_instance":"root","type":"df_complex","type_instance":"free"}]
  | 2018/11/01 20:37:28 receiver.go:127: Debug: Sending message to Notifier channel
  | 2018/11/01 20:37:28 client.go:60: Debug:github.com/satori/go.uuid:   3ae3bc40-5b82-4112-8327-b4e954cc9ea6
  | 2018/11/01 20:37:28 client.go:60: Debug:Printing body {"values":0,"dstypes":"gauge","dsnames":"value","time":1541104650.005,"interval":10.000,"host":"nfvha-strg-03.oot.lab.eng.bos.redhat.com","plugin":"df","plugin_instance":"root","type":"df_complex","type_instance":"reserved"}
  | 2018/11/01 20:37:28 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:28 receiver.go:127: time %!s(int=60)
  | 2018/11/01 20:37:28 receiver.go:127: AMQP Receiving messages.
  | 2018/11/01 20:37:28 client.go:60: Debug:Indexed  3ae3bc40-5b82-4112-8327-b4e954cc9ea6 to index collectd_generic, type event
  | 2018/11/01 20:37:28 client.go:60: Debug:github.com/satori/go.uuid:   9511fffd-595a-4315-8fa7-5db3c487c24d
  | 2018/11/01 20:37:28 client.go:60: Debug:Printing body {"values":49611153408,"dstypes":"gauge","dsnames":"value","time":1541104650.005,"interval":10.000,"host":"nfvha-strg-03.oot.lab.eng.bos.redhat.com","plugin":"df","plugin_instance":"root","type":"df_complex","type_instance":"free"}
  | 2018/11/01 20:37:28 receiver.go:127: Debug: Getting message from AMQP[{"values":[0],"dstypes":["gauge"],"dsnames":["value"],"time":1541104650.005,"interval":10.000,"host":"nfvha-strg-03.oot.lab.eng.bos.redhat.com","plugin":"df","plugin_instance":"root","type":"percent_bytes","type_instance":"reserved"}]
  | 2018/11/01 20:37:28 receiver.go:127: Debug: Sending message to Notifier channel
  | 2018/11/01 20:37:28 receiver.go:127: time %!s(int=60)
  | panic: interface conversion: interface {} is nil, not map[string]interface {}
  |  
  | goroutine 82 [running]:
  | github.com/redhat-nfvpe/telemetry-consumers/internal/pkg/alerts.(*Alerts).Parse(0xc42005ce00, 0xc4203861e0, 0xe8, 0xf0, 0xc42005e150, 0x66)
  | /go/src/github.com/redhat-nfvpe/telemetry-consumers/internal/pkg/alerts/alerts.go:93 +0xa24
  | github.com/redhat-nfvpe/telemetry-consumers/internal/pkg/events.StartEvents.func5(0xc4203842d0, 0xe8, 0xb0c3a0, 0x10, 0xb046e6, 0x5, 0xc420019d70, 0x24)
  | /go/src/github.com/redhat-nfvpe/telemetry-consumers/internal/pkg/events/events.go:219 +0x27b
  | created by github.com/redhat-nfvpe/telemetry-consumers/internal/pkg/events.StartEvents
  | /go/src/github.com/redhat-nfvpe/telemetry-consumers/internal/pkg/events/events.go:215 +0x107c


```

OK I think I can reproduce this now. Basically the code path is resulting in sending events to Alertmanager causing a panic in the Go code. In my case, the reproduction of the issue happens in a valid configuration where events are attempted to be directed to the Alertmanager for delivery.

```
2020/05/05 18:56:20 receiver.go:86: Message ACKed: Message{address: /collectd/cloud1-notify, body: [{"labels":{"alertname":"collectd_connectivity_gauge","instance":"compute-0.redhat.local","connectivity":"tap8f681764-4f","type":"interface_status","severity":"FAILURE","service":"collectd"},"annotations":{"summary":"","ves":"{\"domain\":\"stateChange\",\"eventId\":1,\"eventName\":\"interface tap8f681764-4f down\",\"lastEpochMicrosec\":1588704980836707,\"priority\":\"high\",\"reportingEntityName\":\"collectd connectivity plugin\",\"sequence\":0,\"sourceName\":\"tap8f681764-4f\",\"startEpochMicrosec\":1588704980836629,\"version\":1.0,\"stateChangeFields\":{\"newState\":\"outOfService\",\"oldState\":\"inService\",\"stateChangeFieldsVersion\":1.0,\"stateInterface\":\"tap8f681764-4f\"}}"},"startsAt":"2020-05-05T18:56:20.836640705Z"}]}
2020/05/05 18:56:20 receiver.go:86: Message received... [{"labels":{"alertname":"collectd_connectivity_gauge","instance":"compute-0.redhat.local","connectivity":"tap8f681764-4f","type":"interface_status","severity":"FAILURE","service":"collectd"},"annotations":{"summary":"","ves":"{\"domain\":\"stateChange\",\"eventId\":1,\"eventName\":\"interface tap8f681764-4f down\",\"lastEpochMicrosec\":1588704980836707,\"priority\":\"high\",\"reportingEntityName\":\"collectd connectivity plugin\",\"sequence\":0,\"sourceName\":\"tap8f681764-4f\",\"startEpochMicrosec\":1588704980836629,\"version\":1.0,\"stateChangeFields\":{\"newState\":\"outOfService\",\"oldState\":\"inService\",\"stateChangeFieldsVersion\":1.0,\"stateInterface\":\"tap8f681764-4f\"}}"},"startsAt":"2020-05-05T18:56:20.836640705Z"}]
2020/05/05 18:56:20 client.go:69: Debug:github.com/satori/go.uuid:   f643b2b5-9442-4ea0-8462-3cb3cf09c9ad
2020/05/05 18:56:20 client.go:69: Debug:Printing body map[annotations:map[summary: ves:map[domain:stateChange eventId:%!s(float64=1) eventName:interface tap8f681764-4f down lastEpochMicrosec:%!s(float64=1.588704980836707e+15) priority:high reportingEntityName:collectd connectivity plugin sequence:%!s(float64=0) sourceName:tap8f681764-4f startEpochMicrosec:%!s(float64=1.588704980836629e+15) stateChangeFields:map[newState:outOfService oldState:inService stateChangeFieldsVersion:%!s(float64=1) stateInterface:tap8f681764-4f] version:%!s(float64=1)]] labels:map[alertname:collectd_connectivity_gauge connectivity:tap8f681764-4f instance:compute-0.redhat.local service:collectd severity:FAILURE type:interface_status] startsAt:2020-05-05T18:56:20.836640705Z]
2020/05/05 18:56:20 client.go:69: Debug:Indexed  f643b2b5-9442-4ea0-8462-3cb3cf09c9ad to index collectd_connectivity, type event
2020/05/05 18:56:20 receiver.go:86: Message ACKed: Message{address: /collectd/cloud1-notify, body: [{"labels":{"alertname":"collectd_connectivity_gauge","instance":"compute-0.redhat.local","connectivity":"tap8f681764-4f","type":"interface_status","severity":"OKAY","service":"collectd"},"annotations":{"summary":"","ves":"{\"domain\":\"stateChange\",\"eventId\":2,\"eventName\":\"interface tap8f681764-4f up\",\"lastEpochMicrosec\":1588704980888331,\"priority\":\"high\",\"reportingEntityName\":\"collectd connectivity plugin\",\"sequence\":0,\"sourceName\":\"tap8f681764-4f\",\"startEpochMicrosec\":1588704980888303,\"version\":1.0,\"stateChangeFields\":{\"newState\":\"inService\",\"oldState\":\"outOfService\",\"stateChangeFieldsVersion\":1.0,\"stateInterface\":\"tap8f681764-4f\"}}"},"startsAt":"2020-05-05T18:56:20.888316941Z"}]}
2020/05/05 18:56:20 receiver.go:86: Message received... [{"labels":{"alertname":"collectd_connectivity_gauge","instance":"compute-0.redhat.local","connectivity":"tap8f681764-4f","type":"interface_status","severity":"OKAY","service":"collectd"},"annotations":{"summary":"","ves":"{\"domain\":\"stateChange\",\"eventId\":2,\"eventName\":\"interface tap8f681764-4f up\",\"lastEpochMicrosec\":1588704980888331,\"priority\":\"high\",\"reportingEntityName\":\"collectd connectivity plugin\",\"sequence\":0,\"sourceName\":\"tap8f681764-4f\",\"startEpochMicrosec\":1588704980888303,\"version\":1.0,\"stateChangeFields\":{\"newState\":\"inService\",\"oldState\":\"outOfService\",\"stateChangeFieldsVersion\":1.0,\"stateInterface\":\"tap8f681764-4f\"}}"},"startsAt":"2020-05-05T18:56:20.888316941Z"}]
2020/05/05 18:56:20 receiver.go:86: Message ACKed: Message{address: /collectd/cloud1-notify, body: [{"labels":{"alertname":"collectd_connectivity_gauge","instance":"compute-0.redhat.local","connectivity":"br-int","type":"interface_status","severity":"FAILURE","service":"collectd"},"annotations":{"summary":"","ves":"{\"domain\":\"stateChange\",\"eventId\":3,\"eventName\":\"interface br-int down\",\"lastEpochMicrosec\":1588704980916869,\"priority\":\"high\",\"reportingEntityName\":\"collectd connectivity plugin\",\"sequence\":0,\"sourceName\":\"br-int\",\"startEpochMicrosec\":1588704980916838,\"version\":1.0,\"stateChangeFields\":{\"newState\":\"outOfService\",\"oldState\":\"inService\",\"stateChangeFieldsVersion\":1.0,\"stateInterface\":\"br-int\"}}"},"startsAt":"2020-05-05T18:56:20.916855451Z"}]}
2020/05/05 18:56:21 receiver.go:86: Message received... [{"labels":{"alertname":"collectd_connectivity_gauge","instance":"compute-0.redhat.local","connectivity":"br-int","type":"interface_status","severity":"FAILURE","service":"collectd"},"annotations":{"summary":"","ves":"{\"domain\":\"stateChange\",\"eventId\":3,\"eventName\":\"interface br-int down\",\"lastEpochMicrosec\":1588704980916869,\"priority\":\"high\",\"reportingEntityName\":\"collectd connectivity plugin\",\"sequence\":0,\"sourceName\":\"br-int\",\"startEpochMicrosec\":1588704980916838,\"version\":1.0,\"stateChangeFields\":{\"newState\":\"outOfService\",\"oldState\":\"inService\",\"stateChangeFieldsVersion\":1.0,\"stateInterface\":\"br-int\"}}"},"startsAt":"2020-05-05T18:56:20.916855451Z"}]
panic: interface conversion: interface {} is float64, not string

goroutine 87 [running]:
github.com/infrawatch/smart-gateway/internal/pkg/events/incoming.assimilateMap(0xc000308ed0, 0xc0002d0c88)
	/go/src/github.com/infrawatch/smart-gateway/internal/pkg/events/incoming/collectd.go:107 +0x432
github.com/infrawatch/smart-gateway/internal/pkg/events/incoming.assimilateMap(0xc000308ea0, 0xc0002d0c88)
	/go/src/github.com/infrawatch/smart-gateway/internal/pkg/events/incoming/collectd.go:95 +0x3d1
github.com/infrawatch/smart-gateway/internal/pkg/events/incoming.(*CollectdEvent).GeneratePrometheusAlert(0xc000308e10, 0xc000374090, 0x87, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/infrawatch/smart-gateway/internal/pkg/events/incoming/collectd.go:120 +0x19a
github.com/infrawatch/smart-gateway/internal/pkg/events/incoming.(*CollectdEvent).GeneratePrometheusAlertBody(0xc000308e10, 0xc000374090, 0x87, 0x4, 0x4, 0xc000374090, 0x87, 0xc000301600)
	/go/src/github.com/infrawatch/smart-gateway/internal/pkg/events/incoming/collectd.go:152 +0x5a
github.com/infrawatch/smart-gateway/internal/pkg/events.notifyAlertManager.func1(0xc000028020, 0xc00039e000, 0xc000123c00, 0xc0001f8960, 0x24)
	/go/src/github.com/infrawatch/smart-gateway/internal/pkg/events/events.go:151 +0x207
created by github.com/infrawatch/smart-gateway/internal/pkg/events.notifyAlertManager
	/go/src/github.com/infrawatch/smart-gateway/internal/pkg/events/events.go:148 +0xca
```

--- a/internal/pkg/events/incoming/collectd.go
+++ b/internal/pkg/events/incoming/collectd.go
@@ -88,6 +88,11 @@ func (evt *CollectdEvent) ParseEvent(data string) error {
 
 //assimilateMap recursively saves content of the given map to destination map of strings
 func assimilateMap(theMap map[string]interface{}, destination *map[string]string) {
+	defer func() { //recover from any panic
+		if r := recover(); r != nil {
+			fmt.Println("Recovered in assimilateMap", r)
+		}
+	}()
 	for key, val := range theMap {
 		switch value := val.(type) {
 		case map[string]interface{}:
@@ -102,6 +107,10 @@ func assimilateMap(theMap map[string]interface{}, destination *map[string]string
 				}
 			}
 			(*destination)[key] = strings.Join(aList, ",")
+		case float64,float32:
+			(*destination)[key] = fmt.Sprintf("%f", value)
+		case int:
+			(*destination)[key] = fmt.Sprintf("%d", value)
 		default:
 			// assimilate KV pair
 			(*destination)[key] = value.(string)

--- a/internal/pkg/events/incoming/collectd.go
+++ b/internal/pkg/events/incoming/collectd.go
@@ -107,7 +107,7 @@ func assimilateMap(theMap map[string]interface{}, destination *map[string]string
 				}
 			}
 			(*destination)[key] = strings.Join(aList, ",")
-		case float64,float32:
+		case float64, float32:
 			(*destination)[key] = fmt.Sprintf("%f", value)
 		case int:
 			(*destination)[key] = fmt.Sprintf("%d", value)

--- a/internal/pkg/events/incoming/collectd.go
+++ b/internal/pkg/events/incoming/collectd.go
@@ -90,7 +90,7 @@ func (evt *CollectdEvent) ParseEvent(data string) error {
 func assimilateMap(theMap map[string]interface{}, destination *map[string]string) {
 	defer func() { //recover from any panic
 		if r := recover(); r != nil {
-			fmt.Println("Recovered in assimilateMap", r)
+			log.Printf("Panic:recovered in assimilateMap %v\n", r)
 		}
 	}()
 	for key, val := range theMap {

